Product, Wishlist, Order, Review 기능 추가

jsham0414 opened the pull request:
## 추가 사항
### Product
- 제품 추가
- 제품 정보 조회
  - 상태가 PUBLIC이나 LINK_ONLY인 제품 조회
- 제품 정보 수정
  - 판매자인지 확인 후 수정
- 제품 삭제
  - 판매자인지 확인 후 삭제
- 판매자 상품 조회
  - 업데이트된 날짜순으로 반환
- 제품 이름 조회
  - 필터로 조회된 리스트 반환
- 랭킹 조회
  - 평점 기준 가장 높은 10개의 상품 조회
- 랭킹 업데이트
  - 1시간마다 제품 평점을 업데이트

### Wishlist
- 장바구니 추가
- 장바구니 조회
  - 추가된 날짜순으로 반환
- 장바구니 수정
  - 본인 확인 후 수정
- 장바구니 삭제
  - 본인 확인 후 삭제

### Order
- 상품 구매 시 추가
- 주문 승인, 주문 거절
  - 판매자 확인 후 상태가 PENDING일 경우 수정
- 주문 취소
  - 구매자 확인 후 상태가 PENDING일 경우 수정
- 제품 ID로 들어온 주문 조회
  - 구매일 순으로 반환
- 유저 ID로 들어온 주문 조회
  - 구매일 순으로 반환

### Review
- 리뷰 조회
  - 하나의 리뷰를 반환
- 유저 업데이트된 조회
  - 작성자의 업데이트된 리스트로 반환
- 제품 업데이트된 조회
  - 제품에 쓰인 업데이트된 리스트로 반환
- 업데이트된 작성
  - Order, User당 한 업데이트된 작성 가능
- 업데이트된 수정
  - 본인 확인 후 수정
- 업데이트된 삭제
  - 본인 확인 후 삭제

### Purchase
- 상품 구매
  - Order에 추가 후 장바구니 데이터 삭제

### @EnumCheck
- 날짜순으로 받은 요청에 해당 Enum인지 체크하는 Validator

## 질문하고 싶은 것들

아직 어떤 데이터가 캐싱에 적합한지 모르겠습니다.
토큰을 확인해야 하는 요청이 올 때마다 User 테이블 조회가 발생하는데 User 데이터도 캐싱에 적합할까요?

## 진행하며 생긴 문제점들

@Transaction과 Dirty Checking 그리고 @Encrypt

User 테이블 조회 이후 DB에 평문이 들어가는 현상이 있었습니다.
일어나는 이유는
1. @Transactional을 적용한 메서드에서 사용자 정보를 가져옵니다.
2. @Encrypt 어노테이션을 적용한 필드는 복호화를 합니다.
3. Hibernate에서 사용자 엔티티에 변화가 생긴 걸로 간주합니다. 즉 Dirty Checking에 걸려서 DB에 평문이 들어가버리게 됩니다.

해결한 방법은 복호화 전에 영속성 컨텍스트를 detach했습니다

jsham0414 committed:
Feat: 물품 등록, Enum Validation 추가

판매자 등록, 물품 등록 구현
@EnumCheck과 들어온 String이 올바른 Enum값인지 확인하는 EnumCheckValidator.java 추가

jsham0414 committed:
Feat: 상품 CRUD 추가

상품 추가, 조회, 수정, 삭제 구현
Specification를 사용한 이름으로 검색 구현

jsham0414 committed:
Fix: 메서드명 수정

jsham0414 committed:
Feat: Wishlist CRUD 추가

jsham0414 committed:
Fix: WishlistController.java

jsham0414 committed:
Feat: Order, Review, Purchase 추가

Product
- 상품 랭킹 추가

Order
- 상품 구매시 추가
- 주문 승인, 주문 거절
   - 판매자 확인 후 상태가 PENDING일 경우 수정
- 주문 취소
   - 구매자 확인 후 상태가 PENDING일 경우 수정
- 제품 ID로 들어온 주문 조회
   - 구매일 순으로 반환
- 유저 ID로 들어온 주문 조회
   - 구매일 순으로 반환

Review
- 리뷰 조회
   - 하나의 리뷰를 반환
- 유저 리뷰 조회
   - 작성자의 리뷰를 리스트로 반환
- 제품 리뷰 조회
   - 제품에 쓰여진 리뷰를 리스트로 반환
- 리뷰 작성
   - Order, User당 한 리뷰만 작성 가능
- 리뷰 수정
   - 본인 확인 후 수정
- 리뷰 삭제
   - 본인 확인 후 삭제

Purchase
- 상품 구매
   - Order에 추가 후 장바구니 데이터 삭제

jsham0414 committed:
Fix: 별점 업데이트, Product 삭제 수정

상태가 PUBLIC인 상품만 랭킹에 반영되게 수정
Product 삭제 요청 시 상태를 DELETED로 설정하고 조회가 되지 않게 수정

jsham0414 committed:
Feat: 유저 캐싱 추가, @Encrypt 버그 수정

Redis
- Header에서 추출한 userId가 유효한지 확인하는 과정이 잦다고 판단해 캐싱을 적용

EncryptData.java
- @Encrypt가 적용된 객체에 한해서 영속성 컨텍스트에서 분리시켜 더티 채킹이 일어나지 않게 수정

User
- Json으로 변환하는데 불필요한 UserDetails 필드들을 직렬화 하지 못하게 수정

spring-kotlin-lover commented on api/src/main/java/com/zerobase/commerce/api/order/service/OrderService.java:
```diff
@@ -0,0 +1,174 @@
+    private final UserRepository userRepository;
+    private final ProductRepository productRepository;
+
+    public OrderDto getOrder(HttpHeaders headers, UUID id) {
```
Http 에 대한 규약이 Service 로 영향을 미친 것 같습니다. HttpHeader 는 parameter 로 받는 스펙이 아닌 userId 와 관련된 스펙을 받는 것이 어떨까요?
추가적으로 Controller Layer 의 역할은 무엇이고, Service Layer 의 역할은 무엇일까요?

service 에서 http 라는 통신규약에 강하게 엮어지니, 통신규약을 바꾸게 되면 service 로직을 전체수정해야되는 일이 발생할 수 있습니다.
따라서 tokenAuthenticator 를 controller 에서 호출하고, 넘겨받은 userId 를 service 의 parameter 로 넘기는 것이 layer 책임이 올바르게 지켜졌다고 볼 수 잇을 것 같아요.

jsham0414 replied:
멘토님 의견을 들어보니 비즈니스 로직이 들어가야 할 곳에 통신규약에 엮여버리니 이상하다는 느낌이 듭니다.
수정하겠습니다 !

spring-kotlin-lover commented on api/src/main/java/com/zerobase/commerce/api/order/service/OrderService.java:
```diff
@@ -0,0 +1,174 @@
+
+        order.setStatus(OrderStatus.APPROVED);
+
+        return OrderDto.fromEntity(orderRepository.save(order));
```
JPA dirty check 라는 기능에 대해 제대로 이해하고 계신다면, save 메서드를 호출해줄 필요가 없는데요

추가해주신 이유가 있으신가요?

jsham0414 replied:
명시적으로 해야 하는 줄 알았습니다. 수정하겠습니다.
JPA에 update 메서드가 없는 이유가 궁금했었는데 알게 되었습니다. 감사합니다 !

spring-kotlin-lover commented on api/src/main/java/com/zerobase/commerce/api/purchase/controller/PurchaseController.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/purchase")
```
purchase 라는 API 가 어색해보이긴 합니다. purchase 의 결과로 order 도메인을 만들게 되는데, orderController 에 포함되는게 좋을 것 같아요.

행위라는 것을 url 에 노출하는 것은 많은 모호함이 있습니다. 행위라는 것만 있기 때문에 client 입장에서 이게 어떤 resource 인지도 짐작이 안오고, 다음에 호출할 수 있는 API 가 order 로 연결된다 라는 연결성도 떨어지게 되기 때문입니다.

 POST /order 라는 API 를 통해 order 의 resource 를 반환하는 형태로 가는 API 설계가 올바르게 느껴집니다.

jsham0414 replied:
Order와 Wishlist 사이에서 고민하다가 만들었는데 만들고 나서도 어색하다는 느낌을 받았었습니다.
다시 생각해보니 post와 purchase 둘 다 의미하는 바가 같아서 수정이 필요하다고 생각됩니다.
수정하겠습니다 !

spring-kotlin-lover commented on database/src/main/java/com/zerobase/commerce/database/order/constant/converter/OrderStatusConverter.java:
```diff
@@ -0,0 +1,16 @@
+import com.zerobase.commerce.database.order.constant.OrderStatus;
+import jakarta.persistence.AttributeConverter;
+
+public class OrderStatusConverter implements AttributeConverter<OrderStatus, String> {
```
JPA 에서 제공하는 `@Enumerated(EnumType.STRING)` 을 column 에 붙여주시면 굳이 Enum 을 attributeConverter 로 정의해서 사용해주실 필요는 없습니다!

jsham0414 replied:
@spring-kotlin-lover 악 알겠습니다. 수정하겠습니다.
너무 많은 양의 변경사항을 한꺼번에 PR에 올리다 보니 당시 제가 무엇을 고민했는지도 까먹었었습니다.
리뷰해주신 내용 보니 피드백 받은 것들이 질문 드리고 싶었던 내용도 포함되어 있다는 것을 느꼈습니다.
앞으론 API별로 PR을 올리도록 하겠습니다. 리뷰 감사합니다 !

spring-kotlin-lover reviewed (commented):
하나의 PR 에 너무 많은 변경사항이 있네요 ^^; API 별로 구분해서 나누어서 작업하고 올려주시는 것이 reviewer 입장에서도 보기에 편하답니다.

PR 의 내용이 많으면, 너무 많은 context 가 담겨있어 이해하기가 많이 어려워지게 됩니다.

jsham0414 committed:
Refactor: PR 리뷰 후속 커밋 1

영속화된 엔티티기 때문에 save 메서드 제거
(JPA에 update 메서드가 없는 이유 -> Dirty Checking 때문)
String과 Enum을 변환하는 기능을 @Enumerated로 수정

jsham0414 committed:
Refactor: PR 리뷰 후속 커밋 2

HttpHeaders에서 userId 추출하는 코드를 Controller 계층으로 이동
의미 없는 User 테이블 조회 코드 제거
Order 엔티티에 구매 관련 메서드 추가
파라미터명 통일을 위한 Controller 리팩토링
